### ISSUE TITLE
[Meta] Message Interleaving, User Sessions, Multiple Target Connections, Statefulness

### ISSUE THREAD

**Issue by andrewazores** [rh-jmc-team/container-jfr#153]
The following is a mostly unordered list of architectural questions and concerns. Separate issues will be broken out for each to discuss specific potential solutions.

~~1. Incoming command messages are currently received on each WebSocket connection and then placed into a singular queue for FIFO processing. This means that two different clients connected to the same ContainerJFR instance can easily (accidentally) interfere with each other by doing things like deleting recordings or changing the active target connection.~~

~~2. ContainerJFR only maintains a connection to a single target JVM (or no connection). This is reflected in the API for both command channel messages as well as the HTTP API, where it is assumed that the client is aware of the connection state. ie if the user wants to download a recording named `A` from target `Foo`, they should first ensure that ContainerJFR is connected to `Foo` and then perform an HTTP request for `A`. This ties in with 1, since there is a timing window where another user can change the connected target to `Bar` and cause the original user to either download a different recording named `A` from `Bar, or get a 404 response if `Bar` has no recording named `A`.~~

~~3. Similarly to 2, command messages and HTTP API paths assume the statefulness of the target JVM connection. That is, other than the `connect` message, the client(s) never need to provide a target identifier - ex. the command for starting a recording looks like `start foo 30 template:continuous`, encoding the name, duration, and events of the recording but without *which* target the recording should be started upon.~~

4. All command responses are multicast to all users, and users are expected to build up their model of what the entire "target landscape" looks like from these received responses. There is no API for a user to quickly determine the entire landscape other than issuing a series of commands, ex `is-connected` and then if true, `list` + `list-saved` + `list-event-templates` + `list-event-types` + etc. and listening for all of the responses.

The reality is that ContainerJFR should simply act as a middle-man between the user(s) and the target(s), while maintaining as little of its own state about those users or those targets as possible, since users can (dis)connect at will, and targets may (dis)appear according to various external events (another user kills the target JVM, or reduces the replica count, or changes the deployment configuration, or the target JVM OOMs, etc.). To the user, the existence of RJMX connections to targets should be entirely transparent - the RJMX connection state(s) should not be reflected at all in the API. Any user should, at any time, be able to start/stop/retrieve a recording on any target, without having to first invoke any kind of state-altering command to explicitly establish a target connection. ie there should be a stateless, many-users-to-one-ContainerJFR and one-ContainerJFR-to-many-targets mapping, which ends up looking like a simple stateless many-users-to-many-targets mapping as much as possible.

**Comment by andrewazores** [comment]
@jiekang let me know if you have any more notes to recap our discussion and I'll edit them into the main issue above.

**Comment by jiekang** [comment]
I think that recaps pretty much everything! Nice.

**Comment by andrewazores** [comment]
Points 1, 2, and 3 are now resolved since #161 has been merged. #4 still needs some consideration with extensions to the HTTP API, which can follow along nicely as enhancements after #167 .

**Comment by andrewazores** [comment]
Closing this as the initial goals have generally been reached. Point 4 is still relevant and interesting, but will require some significant work on the frontend first in order to inform what exactly is required of such an endpoint to allow React to populate its initial data with a single request, so I think it's generally out of scope for now.
